src/analysis/ml.py

```diff
@@ -83,44 +80,70 @@ def pca(dataframe: pd.DataFrame, output_png: str, output_file: str, output_coord
     @param output_file: the filename to save the variance explained by each component
     @param output_coord: the filename to save the coordinates of each algorithm
     """
+
     df = dataframe.reset_index(drop=True)
+    column_names = df.head()
+    column_names = [element.split('-')[1] for element in column_names]
     df = df.transpose()  # based on the algorithms rather than the edges
     X = df.values
 
+    min_shape = min(df.shape)
+    if components < 2:
+        raise ValueError(f"components={components} must be greater than or equal to 2 in the config file.")
+    elif components > min_shape:
+        print(f"n_components={components} is not valid. Setting components to {min_shape}.")
+        components = min_shape
+    if not isinstance(labels, bool):
+        raise ValueError(f"labels={labels} must be True or False")
+
     scaler = StandardScaler()
     scaler.fit(X)  # calc mean and standard deviation
     X_scaled = scaler.transform(X)
 
     # choosing the PCA
-    pca_2 = PCA(n_components=2)
-    pca_2.fit(X_scaled)
-    X_pca_2 = pca_2.transform(X_scaled)
-    variance = pca_2.explained_variance_ratio_ * 100
+    pca = PCA(n_components=components)
+    pca.fit(X_scaled)
+    X_pca = pca.transform(X_scaled)
+    variance = pca.explained_variance_ratio_ * 100
 
     # making the plot
     plt.figure(figsize=(10, 7))
-    sns.scatterplot(x=X_pca_2[:, 0], y=X_pca_2[:, 1], s=70)
+    sns.scatterplot(x=X_pca[:, 0], y=X_pca[:, 1], s=70, hue=column_names, legend=True);
     plt.title("PCA")
     plt.xlabel(f"PC1 ({variance[0]:.1f}% variance)")
     plt.ylabel(f"PC2 ({variance[1]:.1f}% variance)")
 
-    # saving the PCA plot
-    make_required_dirs(output_png)
-    plt.savefig(output_png)
+    # saving the coordinates of each algorithm
```

Comment:
To organize the code, I would move this after the graph labeling so that all of the plotting and plot saving code is together.